benchrun/python/README.md

```diff
@@ -0,0 +1,149 @@
+scalars. `CaseList` will populate a `case_list` attribute` which contains the grid of
+specified cases to be run:
+
+```python
+from benchrun import CaseList
+
+case_list = CaseList(params={"x": [1, 2], "y": ["a", "b", "c"]})
+case_list.case_list
+#> [{'x': 1, 'y': 'a'},
+#>  {'x': 1, 'y': 'b'},
+#>  {'x': 1, 'y': 'c'},
+#>  {'x': 2, 'y': 'a'},
+#>  {'x': 2, 'y': 'b'},
+#>  {'x': 2, 'y': 'c'}]
+```
+
+`CaseList` contains an overridable `filter_cases()` method that can be used to remove
+invalid combinations of parameters, e.g. if an `x` of `2` with a `y` of `b` is not
+viable:
+
+```python
+class MyCaseList(CaseList):
+    def filter_cases(self, case_list: List[Dict[str, Any]]) -> List[Dict[str, Any]]:
+        filtered_case_list = []
+        for case in case_list:
+            if not (case["x"] == 2 and case["y"] == "b"):
+                filtered_case_list.append(case)
+
+        return filtered_case_list
+
+my_case_list = MyCaseList(params={"x": [1, 2], "y": ["a", "b", "c"]})
+my_case_list.case_list
+#> [{'x': 1, 'y': 'a'},
+#>  {'x': 1, 'y': 'b'},
+#>  {'x': 1, 'y': 'c'},
+#>  {'x': 2, 'y': 'a'},
+#>  {'x': 2, 'y': 'c'}]
+```
+
+If there are so many restrictions that it is simpler to specify which cases are
+viable than which are not, the `case_list` parameter of `filter_cases()` can be
+completely ignored and a manually-generated list can be returned.
+
+### `Benchmark`
+
+A `Benchmark` in {benchrun} consists of an `Iteration` instance, a `CaseList`
+instance, and potentially a bit more metadata about how to run it like whether to
+drop disk caches beforehand.
+
+```python
+my_benchmark = Benchmark(iteration=my_iteration, case_list=my_case_list)
+```
+
+This class has a `run()` method to run all cases, or `run_case()` to run a single
+case.
+
+### `BenchmarkList`
+
+A `BenchmarkList` is a lightweight class to tie together all the instances of
+`Benchmark` that should be run together (e.g. all the benchmarks for a package).
+
+```python
+from benchrun import BenchmarkList
+
+my_benchmark_list = BenchmarkList(benchmarks = [my_benchmark])
+```
+
+The class has a `__call__()` method that will run all benchmarks in its list,
+taking care that they all use the same `run_id` so they will all appear together
+on conbench.
+
+## Running benchmarks and sending results to conbench
+
+`BenchmarkList` is designed to work seamlessly with
+{[benchadapt](https://github.com/conbench/conbench/tree/main/benchadapt/python)}'s
+`CallableAdapter` class:
+
+```python
+from benchadapt.adapters import CallableAdapter
+
+my_adapter = CallableAdapter(callable=my_benchmark_list)
+```
+
+Like all adapters, it then has a `run()` method to run all the benchmarks it
+contains (handling generic metadata appropriately for you), a `post_results()`
+method that will send the results to a conbench server, and a `__call__()` method
+that will do both. These are the methods that should be called in whatever CI or
+automated build system will be used for running benchmarks.
```

Comment:
Maybe a quick example of doing this in CI here?

Reply:
I'm not precisely sure what this will look like in CI tbh; presumably for Arrow we'll have to overhaul a bit of arrow-benchmarks-ci. The closest we've come so far was [the Go benchmarks](https://github.com/zeroshade/arrow/blob/go-benchmarks/ci/scripts/go_bench_adapt.py), but that's implementing a new adapter, which doesn't need to happen here. All that really needs to happen is inserting runtime information like `run_reason`. So probably a small Python script which can be called from gh actions or whatnot, but we haven't really figured out best practices for that yet.